channeldb/reports.go

```diff
@@ -0,0 +1,288 @@
+package channeldb
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+var (
+	// reportBucket is a bucket which holds a record of the final set
+	// of resolutions which were processed for a channel.
+	// [reportBucket]
+	//	[chainHashBucket]
+	//		[channelBucket]
+	//			[resolversBucket]
+	reportBucket = []byte("channel-reports")
+
+	// resolversBucket holds the outcome of a channel's resolvers.
+	resolversBucket = []byte("resolvers-bucket")
+)
+
+var (
+	// ErrNoReportBucket is returned if the report bucket has not been
+	// created yet.
+	ErrNoReportBucket = errors.New("no reports bucket present")
+
+	// ErrNoChainHashBucket is returned when we have not created a bucket
+	// for the current chain hash.
+	ErrNoChainHashBucket = errors.New("no chain hash bucket")
+
+	// ErrNoChannel is returned when a channel is not found in the chain
+	// hash bucket.
+	ErrNoChannel = errors.New("channel bucket not found")
+)
+
+// ResolverOutcome describes the outcome for the resolver that the contract
+// court reached. This state is not necessarily final, htlcs on our own
+// commitment are resolved across two resolvers.
+type ResolverOutcome uint8
+
+const (
+	resolverOutcomeUnknown ResolverOutcome = iota
+)
+
+// ResolverReport provides an account of the outcome of a resolver. This differs
+// from a ContractReport because it does not necessarily fully resolve the
+// contract; each step of two stage htlc resolution is included.
+type ResolverReport struct {
+	// OutPoint is the outpoint that was spent by this resolver.
+	OutPoint wire.OutPoint
+
```

Comment:
Perhaps we should also include chain fee or fee rate information?

Reply:
Also perhaps more information w.r.t the actual _type_ of the resolver would be useful? So like timeout vs success, etc.

Reply:
> w.r.t the actual type of the resolver would be useful?

I think that the `ResolverOutcome` tells us exactly what type of resolver it was? Eg, a `IncomingHtlcSuccessTx` outcome can only come from an incoming contest resolver. This does mean that we have somewhat duplicate outcomes - eg `IncomingHTLCClaim`, `OutgoingHTLCClaim`; an alternative approach would be to have more generic types eg `HtlcClaim` and pair them with a resolver type. I like that these types are independent of resolver type (future proof to theoretical changes to resolver structure) but I'm ok with either.

Reply:
I would also think that it is useful to have a field here that describes the commit output type: to_local, incoming_htlc, outgoing_htlc or anchor.

Then the htlc resolvers can use one or two transactions. It is the question how that can best be represented here. Also, what to do with chain fees paid (on the first and possibly second tx), who do they reflect in `Amount`.

I think it is definitely good to unify this report with what we report in `ChannelArbitrator.Report`.

Reply:
To add fees here, I think we'd need up update `SpendDetail` notifications to provide us with more info about the spending tx, since we don't have it on hand in some cases. 

> is useful to have a field here that describes the commit output type

Isn't the output type always apparent from the outcome enum? 

> Then the htlc resolvers can use one or two transactions. It is the question how that can best be represented here.

I think that db level, the two stage resolution is best represented as two separate reports.. We could have a type field which tells us to write/read 2x reports from one key, but I see that combination as something we can do rpc level. The benefit of combining the two at db level isn't clear to me.